File: android/app/src/main/java/dev/yog/Update.java
package dev.yog;

import android.app.Activity;
import android.content.ActivityNotFoundException;
import android.content.Intent;
import android.net.Uri;
import androidx.core.content.FileProvider;
import java.io.File;
import java.util.concurrent.atomic.AtomicBoolean;

/**
 * The release channel's device half (bl-7a68, DESIGN §20): read the newest
 * published release, and — when a person taps — put the system installer in
 * front of them.
 *
 * <h2>Three doors, all polled, none blocking</h2>
 *
 * Both reads answer from a volatile field a worker thread fills, so the frame
 * that calls them never waits on a socket. {@link #state} starts the one feed
 * fetch this process makes, on its first call; {@link #install} starts the one
 * download; {@link #said} is how the row that fired it learns what happened,
 * seconds later, without this class holding a callback into Rust.
 *
 * <h2>Nothing here decides anything</h2>
 *
 * The version comparison, the asset pick and the https requirement are
 * {@code crate::update}'s, under the 100% floor, because this side has no test
 * harness at all. What crosses is the feed body verbatim in the two-line
 * protocol every bridge in this app answers in.
 *
 * <h2>Android will not let a process replace itself</h2>
 *
 * An unattended install needs a privileged installer, a device owner or an app
 * store, and this is none of them. So the last rung is {@code ACTION_VIEW} on
 * the downloaded APK: the platform's own installer, in front of a person, who
 * taps. That is the whole difference between this reconciler and the fleet's
 * other four.
 */
public final class Update {
    private Update() {}

    /** Where the {@code updates/} cache directory is served from, per the manifest. */
    private static final String AUTHORITY = ".updates";

    /** The one MIME type the platform's package installer answers to. */
    private static final String APK = "application/vnd.android.package-archive";

    /** The feed answer, in the two-line protocol, or null while nothing has come back. */
    private static volatile String feed;

    /**
     * Whether the one fetch this process makes has been started, and whether a
     * download is running.
     *
     * <p>Atomics rather than a {@code synchronized} method, and it is not a
     * preference: {@code make apk} pins every JNI name this crate resolves
     * against the dex that carries it (DESIGN §15.7), and it looks for access
     * flags reading exactly {@code PUBLIC STATIC}. A {@code synchronized}
     * modifier adds {@code DECLARED_SYNCHRONIZED} and the entry point stops
     * being one the gate can see. A compare-and-set says the same thing about
     * the one race there is — two frames entering at once — without touching
     * the method's signature.
     */
    private static final AtomicBoolean asked = new AtomicBoolean();

    /** What the download last said, or empty when it has nothing to say. */
    private static volatile String said = "";

    /** Whether a download is running, so a second tap does not start a second one. */
    private static final AtomicBoolean fetching = new AtomicBoolean();

    /**
     * The newest release, as the feed answered it — and the call that starts
     * the asking.
     *
     * <p>Once per process, on the first frame that paints the roster. A phone
     * has no timer worth the battery and this app has no business holding one
     * for a question whose answer changes on the order of days; opening the app
     * IS the cadence, which is also the only moment an offer could be seen.
     */
    public static String state() {
        if (asked.compareAndSet(false, true)) {
            worker(
                    () -> {
                        String read = Release.latest();
                        feed = read;
                    });
        }
        // **Empty is "nothing yet", and it is a THIRD state on purpose.** A
        // refusal spelled here while the socket is still open would be
        // indistinguishable from one the network actually gave, and the Rust
        // side stops asking the moment an answer exists — so a phone with no
        // network would then cache the wrong verdict, and one with a slow one
        // would be asked across the JNI boundary on every frame forever.
        String held = feed;
        return held == null ? "" : held;
    }

    /** What the download last said, for the row that fired it. Empty is silence. */
    public static String said() {
        return said;
    }

    /**
     * Fetch {@code url} and hand it to the system installer.
     *
     * <p>Answered immediately with the sentence {@link #said} will keep
     * repeating, because the download is seconds long and the caller is a
     * frame. A second tap while one is running is the same sentence again and
     * not a second download.
     */
    public static String install(String url) {
        if (App.context() == null) {
            said = App.NO_CONTEXT;
            return said;
        }
        if (!fetching.compareAndSet(false, true)) {
            return said;
        }
        said = "downloading the update…";
        worker(
                () -> {
                    String outcome = fetch(url);
                    said = outcome;
                    fetching.set(false);
                });
        return said;
    }

    /** Download the APK and open the installer on it; the sentence is the outcome. */
    private static String fetch(String url) {
        File apk = Release.download(App.context(), url);
        if (apk == null) {
            return "the update did not download: check this device's network and try again.";
        }
        Activity front = App.front();
        if (front == null) {
            return "yog left the screen before the installer could open: tap update again.";
        }
        Uri uri =
                FileProvider.getUriForFile(front, front.getPackageName() + AUTHORITY, apk);
        Intent intent =
                new Intent(Intent.ACTION_VIEW)
                        .setDataAndType(uri, APK)
                        .addFlags(
                                Intent.FLAG_GRANT_READ_URI_PERMISSION
                                        | Intent.FLAG_ACTIVITY_NEW_TASK);
        try {
            front.startActivity(intent);
        } catch (ActivityNotFoundException e) {
            return "this device has no package installer to hand the update to.";
        }
        return "the installer has the update — tap install, then open yog again.";
    }

    /** One named background thread. Both doors start work the frame may not wait on. */
    private static void worker(Runnable work) {
        Thread thread = new Thread(work, "yog-update");
        thread.setDaemon(true);
        thread.start();
    }
}

File: src/update.rs
//! **The phone's release channel, read half** (bl-7a68, DESIGN §20): is the
//! newest published APK newer than this one, and where is it.
//!
//! Every other box in the fleet reconciles itself — the engine off `ghcr.io`,
//! a workstation off the crates.io sparse index, the seat and the foot off
//! their own crates. All five ask one question: *read the newest live version
//! off a registry, compare it with what is installed, install if it differs.*
//! **Android will not let a process replace itself**, so this device answers
//! the first two rungs and stops at the third: it detects, and it OFFERS.
//! The install is the system installer's, in front of a person who taps.
//!
//! This module is the whole of the decision and it is pure: a string in — the
//! `dev.yog.Update` bridge's two-line answer, carrying the body of `GET
//! /repos/<owner>/<repo>/releases/latest` — and an [`Offer`] or nothing out.
//! The platform is on the other side of that string, which is what puts the
//! version comparison and the asset pick under the 100% floor rather than in
//! Java where nothing here can test them.
//!
//! **What it will not do.** No downgrade: the tag has to be strictly greater,
//! so deleting a release and publishing another is the only rollback there is
//! (there is no yank for a GitHub Release, and this states that rather than
//! implying otherwise). No version list: `latest` is one release, exactly as
//! the other four reconcilers read one. And no offer at all from a plaintext
//! URL — an APK is code this device is about to run, so the one transport it
//! may arrive over is the one the feed itself came over.

/// What is on offer: the version the newest release names, and the APK asset
/// under it. Owned and concrete — the caller hands the URL back to the bridge
/// a frame or a minute later, so a borrow of the feed would be a borrow of a
/// string the JNI call already dropped.
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Offer {
    pub version: String,
    pub url: String,
}

/// The two-line protocol's success prefix, the same one every bridge in this
/// app answers in (`dev.yog.App.OK`).
const OK: &str = "ok\n";

/// The one file extension an offer may name. The release workflow attaches
/// exactly one APK and nothing else that ends this way, and a release with no
/// APK at all — a source-only tag, a half-finished upload — is an offer this
/// app must not make rather than a failure worth a sentence.
const APK: &str = ".apk";

/// **Has the feed concluded?** The bridge answers nothing at all — the empty
/// string — while its one fetch is still in flight, and a two-line answer once
/// it has finished, whether that is the release document or a refusal. One
/// caller needs those apart: the shell asks once per frame until this is true
/// and then never again, because a fetch that has concluded cannot conclude
/// differently.
///
/// **The silence is a third state and not a spelling choice.** A refusal
/// written while the socket was still open would be indistinguishable from one
/// the network actually gave, so a phone with no network would cache the wrong
/// verdict and a phone with a slow one would be crossed the JNI boundary on
/// every frame forever.
pub fn answered(answer: &str) -> bool {
    !answer.is_empty()
}

/// Read the bridge's answer against the version this build IS.
///
/// `None` is every state that offers nothing, and they are deliberately one
/// answer: the fetch has not come back, the network refused, the repository
/// has no release yet, the body was not JSON, the tag is not newer, the
/// release carries no APK. A phone with nothing to say about updates says
/// nothing — the offer is one row on a screen, never a nag, so the absence of
/// a row IS the vocabulary for all six.
pub fn offer(answer: &str, installed: &str) -> Option<Offer> {
    let body = answer.strip_prefix(OK)?;
    let feed: serde_json::Value = serde_json::from_str(body).ok()?;
    let tag = feed.get("tag_name")?.as_str()?;
    let version = tag.strip_prefix('v').unwrap_or(tag);
    if !newer(installed, version) {
        return None;
    }
    let url = asset(feed.get("assets")?.as_array()?)?;
    Some(Offer {
        version: version.to_owned(),
        url,
    })
}

/// The APK's download URL out of a release's asset list — the first asset
/// whose name ends in `.apk` and whose URL is `https`.
///
/// The scheme is checked here rather than trusted from the host, because what
/// this URL reaches is handed to the system installer: a feed that named
/// `http://` would be a downgrade of the channel's transport arranged by
/// whoever answered the read.
fn asset(assets: &[serde_json::Value]) -> Option<String> {
    assets
        .iter()
        .filter(|asset| {
            asset
                .get("name")
                .and_then(serde_json::Value::as_str)
                .is_some_and(|name| name.ends_with(APK))
        })
        .find_map(|asset| {
            let url = asset.get("browser_download_url")?.as_str()?;
            url.starts_with("https://").then(|| url.to_owned())
        })
}

/// Is `latest` a strictly greater version than `installed`?
///
/// Dotted decimal, component by component, a missing component reading zero
/// so `0.1` and `0.1.0` are the same version rather than an ordering nobody
/// intended. **Anything that does not parse as decimal is not newer** — a
/// pre-release suffix, a tag somebody typed by hand, an empty string. That is
/// fail-closed in the direction that matters: the failure of a version
/// comparison must be *offer nothing*, never *offer whatever this was*.
pub fn newer(installed: &str, latest: &str) -> bool {
    let (Some(was), Some(is)) = (parts(installed), parts(latest)) else {
        return false;
    };
    // Padded to one width first, so the comparison is a plain lexicographic
    // one over equal-length vectors — `0.1` against `0.1.0` is then equality
    // rather than a length rule written out beside the ordering.
    let width = was.len().max(is.len());
    padded(&is, width) > padded(&was, width)
}

/// One version's components, zero-extended to `width`.
fn padded(version: &[u64], width: usize) -> Vec<u64> {
    (0..width)
        .map(|at| version.get(at).copied().unwrap_or(0))
        .collect()
}

/// One version as its decimal components, or `None` when any of them is not
/// a decimal. The whole version is refused rather than the bad component
/// skipped: a version with an unreadable part is one this app cannot order,
/// and ordering it anyway is how a `0.1.0-rc1` becomes an offer.
fn parts(version: &str) -> Option<Vec<u64>> {
    let version = version.strip_prefix('v').unwrap_or(version);
    if version.is_empty() {
        return None;
    }
    version.split('.').map(|part| part.parse().ok()).collect()
}

#[cfg(test)]
mod tests;
